File: [1277][Medium][Count_Square_Submatrices_with_All_Ones]/Count_Square_Submatrices_with_All_Ones.py
# leetcode time     cost : 736 ms
# leetcode memory   cost : 15.4 MB 
# Time  Complexity: O(MN)
# Space Complexity: O(MN)
# solution 1, DP, refer to No.221
class Solution:
    def countSquares(self, matrix: [[int]]) -> int:
        m, n = len(matrix), len(matrix[0])
        f = [[0] * n for _ in range(m)]
        ans = 0
        for i in range(m):
            for j in range(n):
                if i == 0 or j == 0:
                    f[i][j] = matrix[i][j]
                elif matrix[i][j] == 0:
                    f[i][j] = 0
                else:
                    f[i][j] = min(f[i][j - 1], f[i - 1][j], f[i - 1][j - 1]) + 1
                ans += f[i][j]
        return ans

def main():
    matrix = [[0,1,1,1],[1,1,1,1],[0,1,1,1]]  # expect is 15
    Solution_obj = Solution()
    result = Solution_obj.countSquares(matrix)
    print("result value is ",result)
    
if __name__ =='__main__':
    main()  

File: [1277][Medium][Count_Square_Submatrices_with_All_Ones]/Count_Square_Submatrices_with_All_Ones.cpp
// leetcode time     cost : 88 ms
// leetcode memory   cost : 20.1 MB 
// Time  Complexity: O(MN)
// Space Complexity: O(1)
// solution 1, DP, refer to No.221
class Solution {
public:
    int countSquares(vector<vector<int>>& A) {
        int res = 0;
        for (int i = 0; i < A.size(); ++i)
            for (int j = 0; j < A[0].size(); res += A[i][j++])
                if (A[i][j] && i && j)
                    A[i][j] += min({A[i - 1][j - 1], A[i - 1][j], A[i][j - 1]});
        return res;
    }
};